docs/how-it-works/internals.md

```diff
@@ -0,0 +1,120 @@
+# Internal Architecture
+
+> Last edited April 2024.
+
+This document will outline Lonboard's internal architecture, and hopefully should be useful for potential contributors.
+
+## Pydeck
+
+Before we consider Lonboard, we should briefly cover how Pydeck manages data transit.
+
+<!-- It's impossible to consider Lonboard's architecture in isolation because many of Lonboard's architectural decisions were chosen in explicit contrast to pydeck. Lonboard and pydeck both bind to deck.gl, but have significantly different APIs and characteristics. -->
+
+Pydeck uses the loaders.gl ecosystem natively supported in deck.gl. This means that pydeck passes the user input directly into the `data` prop of any given layer. When `data` is a string, it's interpreted as a URL, passed into loaders.gl machinery to parse the given data format to a list of GeoJSON features. In order to do data-driven styling, users pass in strings like `"@@=properties.valuePerSqm"` to describe the JavaScript callback that should be applied to each GeoJSON feature.
+
+In my opinion, this presents performance pitfalls and gives Pydeck a stifling amount of flexibility. The Python API isn't able to do much of anything because the user input could be _anything_ and users end up writing a sort of JavaScript in Python strings.
```

Comment:
> > The Python API isn't able to do much of anything
> 
> Not entirely accurate, there is nothing stopping the user transforming the data and passing it in the data prop. The accessor then just references the Python-computed property by key. You can certainly point out perf weaknesses here, but it's not like it is impossible to apply transformations in Python

I'm referring to pydeck's own Python code. Of course the user can do whatever they want before passing data into the Pydeck API. But Pydeck is not able to do anything smart with the user input because there's no constraint on the shape of data the user can pass in. By constraining to GeoPandas GeoDataFrames and Arrow tables, Lonboard's _own code_ is able to do a lot of smart stuff for the user, without them needing to be aware/learn anything.